util/election/runner.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/util"
+)
+
+// Config is a set of parameters for a continuous master election process.
+type Config struct {
+	// PreElectionPause is the maximum interval to wait before starting election.
+	PreElectionPause time.Duration
+	// MasterCheckInterval is the interval between checks that the instance still
+	// holds mastership.
+	MasterCheckInterval time.Duration
+	// MasterTTL is the time after which the instance will resign if encountered
+	// no successful IsMaster calls.
+	MasterTTL time.Duration
+}
+
+// Run contains mastership status of a single instance during a single run of
+// being master.
+type Run struct {
+	// Start is the time when the instance became the master as measured by the
+	// corresponding Runner's TimeSource.
+	Start time.Time
+	// Ctx is the context which is active while the instance believes to be the
+	// master. As soon as Ctx is canceled, the instance attempts to resign, and
+	// stops monitoring its status.
+	Ctx context.Context
+	// Cancel cancels Ctx, or does nothing if that is already canceled.
+	// Effectively, it instructs the instance to resign if not yet.
+	Cancel context.CancelFunc
+	// Done closes when mastership status monitoring goroutine for this run
```

Comment:
Is the completion channel really needed? 

(And if it is, could it be a `sync.WaitGroup`, so if the caller forgets to `run.Resign()` the monitoring goroutine doesn't block before terminating)

Just looking at runner.go, I don't quite follow why `run.Resign()` can't just `me.Resign()` and cancel the inner context, then assume the monitoring goroutine will die off on its own.

Reply:
Currently, if the caller forgets to `run.Resign()` (or even deliberately doesn't run it because they never want to resign), the goroutine will still terminate when the `Context` used for its creation is canceled or mastership is lost. And, if I understand you correctly, replacing the channel by `WaitGroup` would be equivalent w.r.t. the goroutine's lifetime, only the reporting mechanism would differ.

Good idea about `me.Resign()` though, will try this out.